stdlib/public/stubs/UnicodeScalarProps.cpp

```diff
@@ -0,0 +1,673 @@
+  0x961F3F8, 0x1181F3FB, 0x961F400, 0xC81F43F, 0x961F440, 0xC81F441, 0x1541F442, 0x961F444,
+  0x1541F446, 0x961F451, 0x1541F466, 0x961F479, 0x1541F47C, 0x961F47D, 0x1541F481, 0x961F484,
+  0x1541F485, 0x961F488, 0x1541F48F, 0x961F490, 0x1541F491, 0x961F492, 0x1541F4AA, 0x961F4AB,
+  0xC81F4FD, 0x9E1F4FE, 0x961F4FF, 0x9E1F53E, 0xC81F549, 0x961F54B, 0x9E1F54F, 0x961F550, 0x9E1F568,
+  0xC81F56F, 0x9E1F571, 0xC81F573, 0xAC1F574, 0xC81F576, 0x1541F57A, 0x9E1F57B, 0xC81F587,
+  0x9E1F588, 0xC81F58A, 0x9E1F58E, 0xAC1F590, 0x9E1F591, 0x1541F595, 0x9E1F597, 0x961F5A4,
+  0xC81F5A5, 0x9E1F5A6, 0xC81F5A8, 0x9E1F5A9, 0xC81F5B1, 0x9E1F5B3, 0xC81F5BC, 0x9E1F5BD, 0xC81F5C2,
+  0x9E1F5C5, 0xC81F5D1, 0x9E1F5D4, 0xC81F5DC, 0x9E1F5DF, 0xC81F5E1, 0x9E1F5E2, 0xC81F5E3, 0x9E1F5E4,
+  0xC81F5E8, 0x9E1F5E9, 0xC81F5EF, 0x9E1F5F0, 0xC81F5F3, 0x9E1F5F4, 0xC81F5FA, 0x961F5FB,
+  0x1541F645, 0x961F648, 0x1541F64B, 0x9E1F650, 0x961F680, 0x1541F6A3, 0x961F6A4, 0x1541F6B4,
+  0x961F6B7, 0x1541F6C0, 0x961F6C1, 0x9E1F6C6, 0xC81F6CB, 0x1541F6CC, 0xC81F6CD, 0x961F6D0,
+  0x9E1F6D3, 0x961F6D5, 0xB01F6D8, 0x961F6DD, 0xC81F6E0, 0x9E1F6E6, 0xC81F6E9, 0x9E1F6EA, 0x961F6EB,
+  0xB01F6ED, 0xC81F6F0, 0x9E1F6F1, 0xC81F6F3, 0x961F6F4, 0xB01F6FD, 0x9E1F700, 0xB01F774, 0x9E1F780,
+  0xB01F7D9, 0x961F7E0, 0xB01F7EC, 0x961F7F0, 0xB01F7F1, 0x9E1F800, 0xB01F80C, 0x9E1F810, 0xB01F848,
+  0x9E1F850, 0xB01F85A, 0x9E1F860, 0xB01F888, 0x9E1F890, 0xB01F8AE, 0x9E1F8B0, 0xB01F8B2, 0x9E1F900,
+  0x1541F90C, 0x961F90D, 0x1541F90F, 0x961F910, 0x1541F918, 0x961F920, 0x1541F926, 0x961F927,
+  0x1541F930, 0x961F93A, 0x9E1F93B, 0x1541F93C, 0x961F93F, 0x9E1F946, 0x961F947, 0x1541F977,
+  0x961F978, 0x1541F9B5, 0x961F9B7, 0x1541F9B8, 0x961F9BA, 0x1541F9BB, 0x961F9BC, 0x1541F9CD,
+  0x961F9D0, 0x1541F9D1, 0x961F9DE, 0x9E1FA00, 0xB01FA54, 0x9E1FA60, 0xB01FA6E, 0x961FA70,
+  0xB01FA75, 0x961FA78, 0xB01FA7D, 0x961FA80, 0xB01FA87, 0x961FA90, 0xB01FAAD, 0x961FAB0, 0xB01FABB,
+  0x961FAC0, 0x1541FAC3, 0xB01FAC6, 0x961FAD0, 0xB01FADA, 0x961FAE0, 0xB01FAE8, 0x1541FAF0,
+  0xB01FAF7, 0x9E1FB00, 0xB01FB93, 0x9E1FB94, 0xB01FBCB, 0x841FBF0, 0xB01FBFA, 0x121FFFE, 0x4020000,
+  0xB02A6E0, 0x402A700, 0xB02B739, 0x402B740, 0xB02B81E, 0x402B820, 0xB02CEA2, 0x402CEB0, 0xB02EBE1,
+  0x562F800, 0xB02FA1E, 0x122FFFE, 0x4030000, 0xB03134B, 0x123FFFE, 0xB040000, 0x124FFFE, 0xB050000,
+  0x125FFFE, 0xB060000, 0x126FFFE, 0xB070000, 0x127FFFE, 0xB080000, 0x128FFFE, 0xB090000, 0x129FFFE,
+  0xB0A0000, 0x12AFFFE, 0xB0B0000, 0x12BFFFE, 0xB0C0000, 0x12CFFFE, 0xB0D0000, 0x12DFFFE,
+  0x122E0000, 0x36E0001, 0x122E0002, 0x162E0020, 0x122E0080, 0x1AE0100, 0x122E01F0, 0xB0E1000,
+  0x12EFFFE, 0xB0F0000, 0x12FFFFE, 0xB100000, 0x130FFFE,
+};
+
+SWIFT_RUNTIME_STDLIB_INTERNAL
+__swift_bool _swift_stdlib_hasBinaryProperty(__swift_uint32_t scalar,
+                                             __swift_intptr_t propertyMask) {
+
+  __swift_uint64_t binaryProperties = 0;
+
+  auto low = 0;
+  auto high = 4855 - 1;
```

Comment:
Is this all the hashing stuff? Any way to share helper code with normalization and grapheme breaking, or is this a completely different path?

Reply:
Are you referring to the `auto high = 4855 - 1;` or the function in general? If just the high variable, that's just the size of the array - 1 (max index). If you're referring to the function in general, it is just a binary search over the array, but I have tried looking into defining a helper function for binary search. There were too many differences between each of them that it made it difficult, but I can give another go at it.

Reply:
If we're writing this code, we should definitely use names/macros for magic numbers. If we're generating the code, then we really have no excuse :-).